### PR TITLE
Added TOP5 tab

### PR DESCRIPTION
**Pull request by albert-nowak**
**JIRA ticket:** [WDP-140](https://kindergarten.stxnext.pl/browse/WDP-140)

### CODE REVIEW

File: src/presence_analyzer/tests.py [hunk 1]
```diff
@@ -155,6 +155,57 @@ def test_presence_start_end_view(self):
...
+        resp = self.client.get('/api/v1/top_5/September-2013')
+        self.assertEqual(resp.status_code, 200)
+        data = json.loads(resp.data)
+        self.assertEqual(
+            data[0],
+            [
+                11,
+                {
+                    u'total_hours': 32.89,
+                    u'image': '<img src="https://intranet.stxnext.pl/api/images/users/11"/>',
```
Comment (albert-nowak) [inline src/presence_analyzer/tests.py]: This line is too long but I don't know how to break it correctly

File: src/presence_analyzer/tests.py [hunk 2]
```diff
@@ -155,6 +155,57 @@ def test_presence_start_end_view(self):
...
+
+    def test_api_months_view(self):
+        """
+        Test months listing.
+        """
+        resp = self.client.get('/api/v1/months')
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        data = json.loads(resp.data)
+        self.assertEqual(data[0], u'September-2013')
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: u'?

File: src/presence_analyzer/views.py [hunk 3]
```diff
@@ -75,6 +85,62 @@ def users_view():
     ]
 
 
+@app.route('/api/v1/months', methods=['GET'])
+@jsonify
+def months_view():
+    """
+    Years/months listing for dropdown.
+    """
+    months = get_dates()
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: return get_dates()

File: src/presence_analyzer/views.py [hunk 4]
```diff
@@ -75,6 +86,59 @@ def users_view():
...
+    """
+    Returns top 5 employees of given month.
+    """
+    data = get_data()
+    top_presence = OrderedDict()
+
+    month_year = month_year.split('-')
+    try:
+        month = month_year[0]
+        year = month_year[1]
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: why not
`month, year = month_year.split('-')`

File: src/presence_analyzer/views.py [hunk 5]
```diff
@@ -75,6 +86,59 @@ def users_view():
...
+    except IndexError:
+        log.debug('{} is not correct format'.format(month_year))
+        abort(404)
+
+    if (len(year) < 4 or
+            not year.isdigit() or
+            month not in month_name):
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: style

File: src/presence_analyzer/templates/presence_top_5.html [hunk 6]
```diff
@@ -0,0 +1,61 @@
+<%inherit file="base.html"/>
+
+<%block name="title">
+    <title>TOP 5</title>
+</%block>
+
+<%block name="head">
+    <script type="text/javascript">
+        (function($) {
+            $(document).ready(function() {
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_top_5.html]: `$(function() {` is shorter

File: src/presence_analyzer/templates/presence_top_5.html [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                var $loading = $('#loading'),
+                    xhr = null;
+                $.getJSON("/api/v1/months", function(result) {
+                    var dropdown = $("#object_id");
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_top_5.html]: singleqotes

File: src/presence_analyzer/templates/presence_top_5.html [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                    $.each(result, function(item) {
+                        dropdown.append($("<option />").val(result[item]).text(result[item]));
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_top_5.html]: ditto
and this imo is more readable

    
    $('<option />', {
        'val': result[item],
        'text': result[item]
    })

File: src/presence_analyzer/templates/presence_top_5.html [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+                    });
+                    dropdown.show();
+                    $loading.hide();
+                });
+                $('#object_id').change(function() {
+                    var selectedMonth = $("#object_id").val(),
+                        $chartDiv = $('#chart_div');
+                    if (xhr) {
+                            xhr.abort();
+                            xhr = null;
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_top_5.html]: indent

File: src/presence_analyzer/utils.py [hunk 10]
```diff
@@ -216,3 +244,17 @@ def work_hours(items):
...
+def total_hours(items, month, year):
+    """
+    Returns total working hours.
+    """
+    seconds = 0
+    if not items:
+        return 0
+    for date in items:
+        if (calendar.month_name[date.month] == month and
+                int(date.year) == int(year)):
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: style

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/views.py [hunk 11]
```diff
@@ -75,6 +86,59 @@ def users_view():
...
+        log.debug('{} is not correct format'.format(month_year))
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: In logging don't use str.format, use `%`
`log.debug('%s xxxx %s', var1, var2)`

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/views.py [hunk 12]
```diff
@@ -75,6 +86,57 @@ def users_view():
...
+        sorted(
+            top_presence.items(),
+            key=lambda x: x[1]['total_hours'],
+            reverse=True
+        )
+    ).items())[:5]
+    if not top5[0][1]['total_hours']:
+        log.debug('No data for year %s', year)
+        abort(404)
+    else:
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: Is that neccesary?

File: src/presence_analyzer/views.py [hunk 13]
```diff
@@ -75,6 +86,57 @@ def users_view():
...
+    data = get_data()
+    top_presence = OrderedDict()
+
+    try:
+        month, year = month_year.split('-')
+    except IndexError:
+        log.debug('%s is not a correct format!', month_year)
+        abort(404)
+
+    if (len(year) < 4 or
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: Put this in one line.

File: src/presence_analyzer/views.py [hunk 14]
```diff
@@ -75,6 +86,57 @@ def users_view():
...
+        not year.isdigit() or
+        month not in month_name):
+        log.debug('%s is not a correct format!', month_year)
+        abort(404)
+
+    for user in data:
+        top_presence[user] = {
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: use dict comprehension

File: src/presence_analyzer/templates/presence_top_5.html [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+                            data.addColumn('string', 'Name');
+                            data.addColumn('string', 'Image');
+                            data.addColumn('number', 'Hours in work');
+                            data.addRows(list);
+                            $chartDiv.show();
+                            $loading.hide();
+                            table.draw(data, options);
+                        });
+                    }
+                    if (xhr.done()) $dropdown.prop('disabled', false);
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_top_5.html]: never do it in one line
besides
    
    $.getJSON(
        ...
    )
        .done(function () {
            ...
        })

File: src/presence_analyzer/templates/presence_top_5.html [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+                });
+                $('#object_id').change(function() {
+                    $dropdown.prop('disabled', true);
+                    var selectedMonth = $('#object_id').val(),
+                        $chartDiv = $('#chart_div');
+                    if (selectedMonth) {
+                        $loading.show();
+                        $chartDiv.hide();
+                        xhr = $.getJSON("/api/v1/top_5/" + selectedMonth, function(result) {
+                            var data = new google.visualization.DataTable(),
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_top_5.html]: indent

File: src/presence_analyzer/templates/presence_top_5.html [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+                                options = {showRowNumber: true, width: '100%', height: '100%', allowHtml: true},
+                                container = document.getElementById("chart_div"),
+                                table = new google.visualization.Table(container),
+                                list = [];
+                            $.each(result, function(index, value) {
+                                list.push([
+                                    value[1]['name'],
+                                    "<img src=\"" + value[1]['image'] + "\"/>",
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/presence_top_5.html]: `$('', {src: xxx})`

File: src/presence_analyzer/tests.py [hunk 18]
```diff
@@ -155,6 +155,57 @@ def test_presence_start_end_view(self):
             }
         )
 
+    def test_prepare_photo_url(self):
+        """
+        Test preparing correct url for given user_id.
+        """
+        resp = self.client.get('/api/v1/photo_url/abc')
+        self.assertEqual(resp.status_code, 404)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py:163]: enter

File: src/presence_analyzer/tests.py [hunk 19]
```diff
@@ -155,6 +155,57 @@ def test_presence_start_end_view(self):
...
+        resp = self.client.get('/api/v1/photo_url/10')
+        self.assertEqual(resp.status_code, 200)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: enter

File: src/presence_analyzer/utils.py [hunk 20]
```diff
@@ -216,3 +244,17 @@ def work_hours(items):
...
+def total_hours(items, month, year):
+    """
+    Returns total working hours.
+    """
+    seconds = 0
+    if not items:
+        return 0
+    for date in items:
+        if (calendar.month_name[date.month] == month and
+            int(date.year) == int(year)):
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: style

File: src/presence_analyzer/views.py [hunk 21]
```diff
@@ -2,20 +2,23 @@
 """
 Defines views.
 """
-from calendar import day_abbr
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: ?

File: src/presence_analyzer/views.py [hunk 22]
```diff
@@ -55,10 +59,17 @@ def prepare_photo_url(user_id):
     Returns url for intranet api in order to get photo of given user.
     """
     conf = get_server_config()
+
+    try:
+        int(user_id)
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: Why? Check ur route param

File: src/presence_analyzer/views.py [hunk 23]
```diff
@@ -75,6 +86,55 @@ def users_view():
...
+        sorted(
+            top_presence.items(),
+            key=lambda x: x[1]['total_hours'],
+            reverse=True
+        )
+    ).items())[:5]
+    if not top5[0][1]['total_hours']:
+        log.debug('No data for year %s', year)
+        abort(404)
+    return top5
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: Imo there's to much logic here. Such operations should be in some kind of util functions or so. Views should be as light as possible.

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/utils.py (hunk 20 above)
Comment (albert-nowak) [reply src/presence_analyzer/utils.py]: https://www.python.org/dev/peps/pep-0008/#multiline-if-statements
PEP8 says this is a correct indent
`int()` is incorrect?

Comment (dpolcyn) [reply src/presence_analyzer/utils.py]: put this to one line

File: src/presence_analyzer/views.py [hunk 24]
```diff
@@ -75,6 +78,55 @@ def users_view():
...
+    if len(year) < 4 or not year.isdigit() or month not in month_name:
+        log.debug('%s is not a correct format!', month_year)
+        abort(404)
+
+    top_presence = {
+        user: {
+            'total_hours': total_hours(data[user]['presence'], month, year),
+            'image': prepare_photo_url(user).data[1:-1],
+            'name': data[user]['name']
+        } for user in data
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: for in next line

File: src/presence_analyzer/views.py [hunk 25]
```diff
@@ -75,6 +78,55 @@ def users_view():
...
+    }
+
+    top5 = list(OrderedDict(
+        sorted(
+            top_presence.items(),
+            key=lambda x: x[1]['total_hours'],
+            reverse=True
+        )
+    ).items())[:5]
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py:123]: empty line after this

File: src/presence_analyzer/views.py [hunk 26]
```diff
@@ -75,6 +78,55 @@ def users_view():
...
+    return get_dates()
+
+
+@app.route('/api/v1/top_5/<string:month_year>', methods=['GET'])
+@jsonify
+def top_5(month_year):
+    """
+    Returns top 5 employees of given month.
+    """
+    data = get_data()
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: i dont see any reason to get it here

**Pull request merged**